File: backend/src/services/MemberService.cpp
#include "services/MemberService.h"
#include "repositories/MockMemberRepository.h"
#include "repositories/MockUserRepository.h"
#include "services/OperationLogService.h"
#include "services/PermissionService.h"
#include "utils/ApiResponse.h"
#include "utils/ErrorCode.h"
#include "utils/JwtUtil.h"
#include "utils/TimeUtil.h"

namespace smart_home {

static std::string memberJson(const HomeMember& m) {
    return "{\"memberId\":" + std::to_string(m.memberId) + ",\"homeId\":" + std::to_string(m.homeId) +
           ",\"userId\":" + std::to_string(m.userId) + ",\"role\":" + ApiResponse::quote(m.role) +
           ",\"status\":" + ApiResponse::quote(m.status) + ",\"permissionScope\":" + ApiResponse::quote(m.permissionScope) +
           ",\"validTo\":" + ApiResponse::quote(m.validTo) + "}";
}

std::string MemberService::invite(const std::string& token, long long homeId, const std::string& phone, const std::string& role, const std::string& validTo) {
    const auto payload = JwtUtil::parseToken(token);
    if (!payload) return ApiResponse::error(UNAUTHORIZED, "unauthorized");
    if (!PermissionService().canManageHome(payload->userId, homeId)) return ApiResponse::error(FORBIDDEN, "permission denied");
    MockUserRepository userRepo;
    const auto user = userRepo.findByPhone(phone);
    if (!user) return ApiResponse::error(NOT_FOUND, "user not found");
    MockMemberRepository memberRepo;
    const std::string now = TimeUtil::now();
    const auto member = memberRepo.save(HomeMember{0, homeId, user->userId, role.empty() ? "MEMBER" : role, "ACTIVE", "ALL", now, validTo, now, now});
    OperationLogService().write(payload->userId, homeId, "INVITE_MEMBER", "invite member", "MEMBER", member.memberId);
    return ApiResponse::success(memberJson(member));
}

std::string MemberService::apply(const std::string& token, long long homeId, const std::string&) {
    const auto payload = JwtUtil::parseToken(token);
    if (!payload) return ApiResponse::error(UNAUTHORIZED, "unauthorized");
    MockMemberRepository memberRepo;
    const std::string now = TimeUtil::now();
    const auto member = memberRepo.save(HomeMember{0, homeId, payload->userId, "MEMBER", "PENDING", "", now, "", now, now});
    return ApiResponse::success(memberJson(member));
}

std::string MemberService::approve(const std::string& token, long long homeId, long long memberId, bool approved, const std::string& role) {
    const auto payload = JwtUtil::parseToken(token);
    if (!payload) return ApiResponse::error(UNAUTHORIZED, "unauthorized");
    if (!PermissionService().canManageMember(payload->userId, homeId, memberId)) return ApiResponse::error(FORBIDDEN, "permission denied");
    MockMemberRepository repo;
    auto member = repo.findById(memberId);
    if (!member) return ApiResponse::error(NOT_FOUND, "member not found");
    member->status = approved ? "ACTIVE" : "REJECTED";
    if (!role.empty()) member->role = role;
    repo.update(*member);
    OperationLogService().write(payload->userId, homeId, "APPROVE_MEMBER", "approve member", "MEMBER", memberId);
    return ApiResponse::success(memberJson(*member));
}

std::string MemberService::listMembers(const std::string& token, long long homeId) {
    const auto payload = JwtUtil::parseToken(token);
    if (!payload) return ApiResponse::error(UNAUTHORIZED, "unauthorized");
    if (!PermissionService().isMemberOfHome(payload->userId, homeId)) return ApiResponse::error(FORBIDDEN, "permission denied");
    MockMemberRepository repo;
    const auto members = repo.findByHome(homeId);
    std::string data = "[";
    for (size_t i = 0; i < members.size(); ++i) {
        if (i) data += ",";
        data += memberJson(members[i]);
    }
    data += "]";
    return ApiResponse::success(data);
}

std::string MemberService::updatePermission(const std::string& token, long long homeId, long long memberId,
                                            const std::string& role, const std::string& permissionScope, const std::string& validTo) {
    const auto payload = JwtUtil::parseToken(token);
    if (!payload) return ApiResponse::error(UNAUTHORIZED, "unauthorized");
    if (!PermissionService().canManageMember(payload->userId, homeId, memberId)) return ApiResponse::error(FORBIDDEN, "permission denied");
    MockMemberRepository repo;
    auto member = repo.findById(memberId);
    if (!member) return ApiResponse::error(NOT_FOUND, "member not found");
    if (!role.empty()) member->role = role;
    if (!permissionScope.empty()) member->permissionScope = permissionScope;
    if (!validTo.empty()) member->validTo = validTo;
    repo.update(*member);
    OperationLogService().write(payload->userId, homeId, "UPDATE_MEMBER_PERMISSION", "update member permission", "MEMBER", memberId);
    return ApiResponse::success(memberJson(*member));
}

std::string MemberService::removeMember(const std::string& token, long long homeId, long long memberId) {
    const auto payload = JwtUtil::parseToken(token);
    if (!payload) return ApiResponse::error(UNAUTHORIZED, "unauthorized");
    if (!PermissionService().canManageMember(payload->userId, homeId, memberId)) return ApiResponse::error(FORBIDDEN, "permission denied");
    MockMemberRepository repo;
    if (!repo.remove(memberId)) return ApiResponse::error(NOT_FOUND, "member not found");
    OperationLogService().write(payload->userId, homeId, "REMOVE_MEMBER", "remove member", "MEMBER", memberId);
    return ApiResponse::success("{\"status\":\"member_removed\"}");
}

} // namespace smart_home

File: smart_home/smart home/backend/app/services/member_service.py
from app.core.permissions import Permission
from app.repositories.factory import get_member_repository, get_user_repository
from app.services.account_operation_log_service import AccountOperationLogService
from app.services.permission_service import PermissionService
from app.utils.response import AppException


class MemberService:
    def __init__(self):
        self.members = get_member_repository()
        self.users = get_user_repository()
        self.permissions = PermissionService()
        self.logs = AccountOperationLogService()

    def _require_member_manage(self, home_id: int, user_id: int):
        return self.permissions.require_home_permission(home_id, user_id, Permission.MEMBER_MANAGE)

    def invite(self, home_id: int, operator_id: int, phone: str, role: str, expire_at: str | None) -> dict:
        self._require_member_manage(home_id, operator_id)
        existed = self.members.find_by_home_and_phone(home_id, phone)
        if existed and existed.status in {"ACTIVE", "PENDING", "INVITED"}:
            raise AppException(409, "member already exists or pending")
        user = self.users.find_by_phone(phone)
        member = self.members.invite(
            home_id=home_id,
            user_id=user.user_id if user else 0,
            username=user.username if user else "",
            phone=phone,
            role=role,
            expire_at=expire_at,
            invited_by=operator_id,
        )
        self.logs.write(operator_id, home_id, "INVITE_MEMBER", "invite member", "MEMBER", member.member_id)
        return member.__dict__

    def accept_invitation(self, home_id: int, member_id: int, user_id: int) -> dict:
        user = self.users.find_by_id(user_id)
        if not user:
            raise AppException(404, "user not found")
        member = self.members.find_by_id(member_id)
        if not member or member.home_id != home_id:
            raise AppException(404, "invitation not found")
        accepted = self.members.accept_invitation(member_id, user.user_id, user.username, user.phone)
        if not accepted:
            raise AppException(403, "invitation does not belong to current user")
        self.logs.write(user_id, home_id, "ACCEPT_INVITATION", "accept home invitation", "MEMBER", member_id)
        return accepted.__dict__

    def apply(self, home_id: int, user_id: int, reason: str | None) -> dict:
        user = self.users.find_by_id(user_id)
        if not user:
            raise AppException(404, "user not found")
        existed = self.members.find_by_home_and_user(home_id, user_id) or self.members.find_by_home_and_phone(home_id, user.phone)
        if existed and existed.status in {"ACTIVE", "PENDING", "INVITED"}:
            raise AppException(409, "member already exists or pending")
        member = self.members.apply(home_id, user.user_id, user.username, user.phone, reason)
        self.logs.write(user_id, home_id, "APPLY_MEMBER", "apply to join home", "MEMBER", member.member_id)
        return member.__dict__

    def approve(self, home_id: int, operator_id: int, member_id: int, approved: bool, role: str) -> dict:
        self._require_member_manage(home_id, operator_id)
        member = self.members.find_by_id(member_id)
        if not member or member.home_id != home_id:
            raise AppException(404, "member not found")
        updated = self.members.approve(member_id, approved, role)
        if not updated:
            raise AppException(400, "member status cannot be approved")
        self.logs.write(operator_id, home_id, "APPROVE_MEMBER", "approve member", "MEMBER", member_id)
        return updated.__dict__

    def list_members(self, home_id: int, user_id: int) -> list[dict]:
        self.permissions.require_home_permission(home_id, user_id, Permission.HOME_VIEW)
        return [member.__dict__ for member in self.members.list_by_home(home_id)]

    def update_permission(self, home_id: int, operator_id: int, member_id: int, role: str, expire_at: str | None) -> dict:
        self._require_member_manage(home_id, operator_id)
        member = self.members.find_by_id(member_id)
        if not member or member.home_id != home_id:
            raise AppException(404, "member not found")
        updated = self.members.update_permission(member_id, role, expire_at)
        if not updated:
            raise AppException(400, "only active member permission can be updated")
        self.logs.write(operator_id, home_id, "UPDATE_MEMBER_PERMISSION", "update member permission", "MEMBER", member_id)
        return updated.__dict__

    def remove(self, home_id: int, operator_id: int, member_id: int) -> dict:
        self._require_member_manage(home_id, operator_id)
        member = self.members.find_by_id(member_id)
        if not member or member.home_id != home_id:
            raise AppException(404, "member not found")
        if member.user_id == operator_id:
            raise AppException(400, "cannot remove yourself")
        if not self.members.remove(member_id):
            raise AppException(404, "member not found")
        self.logs.write(operator_id, home_id, "REMOVE_MEMBER", "remove member", "MEMBER", member_id)
        return {"status": "removed"}
